tests/src/runtimeApi/stream/StreamAddCallback.cpp

```diff
@@ -0,0 +1,145 @@
+#include <stdio.h>
+#include <hip/hip_runtime.h>
+#include <unistd.h>
+#include "test_common.h"
+#include <atomic>
+
+/* HIT_START
+ * BUILD: %t %s ../../test_common.cpp NVCC_OPTIONS -std=c++11
+ * TEST: %t
+ * HIT_END
+ */
+
+enum class ExecState
+{
+   EXEC_NOT_STARTED,
+   EXEC_STARTED,
+   EXEC_CB_STARTED,
+   EXEC_CB_FINISHED,
+   EXEC_FINISHED
+};
+
+struct UserData
+{
+    size_t size;
+    int* ptr;
+};
+
+// Global variable to check exection order
+std::atomic<ExecState> gData(ExecState::EXEC_NOT_STARTED);
+
+
+void myCallback(hipStream_t stream, hipError_t status, void* user_data)
+{
+    if(gData.load() != ExecState::EXEC_STARTED)
+        return; // Error hence return early
+
+    gData.store(ExecState::EXEC_CB_STARTED);
+
+    UserData* data = reinterpret_cast<UserData*>(user_data);
+    printf("Callback started\n");
+
+    sleep(1);
+
+    printf("Callback ending.\n");
+    gData.store(ExecState::EXEC_CB_FINISHED);
+}
+
+bool test(int count)
+{
+    printf("\n============ Test iteration %d =============\n",count);
+    // Stream
+    hipStream_t stream;
+    bool result = true;
+
+    gData.store(ExecState::EXEC_STARTED);
+
+    HIPCHECK(hipStreamCreate(&stream));
+
+    // Array size
+    size_t size = 10000;
+
+    // Device array
+    int *data = NULL;
+    HIPCHECK(hipMalloc((void**)&data, sizeof(int) * size));
+
+    // Initialize device array to -1
+    HIPCHECK(hipMemset(data, -1, sizeof(int) * size));
+
+    // Host array
+    int *host = NULL;
+    HIPCHECK(hipHostMalloc((void**)&host, sizeof(int) * size));
+
+    // Print host ptr address
+    printf("In main thread\n");
+
+    // Initialize user_data for callback
+    UserData arg;
+    arg.size = size;
+    arg.ptr  = host;
+
+    // Synchronize device
+    HIPCHECK(hipDeviceSynchronize());
+
+    // Asynchronous copy from device to host
+    HIPCHECK(hipMemcpyAsync(host, data, sizeof(int) * size, hipMemcpyDeviceToHost, stream));
+
+    // Asynchronous memset on device
+    HIPCHECK(hipMemsetAsync(data, 0, sizeof(int) * size, stream));
+
+    // Add callback - should happen after hipMemsetAsync()
+    HIPCHECK(hipStreamAddCallback(stream, myCallback, &arg, 0));
+
+    printf("Will wait in main thread until callback completes\n");
+
+    //This should synchronize the stream (including the callback)
+    HIPCHECK(hipStreamSynchronize(stream));
```

Comment:
This is _your_ test program, taken from https://github.com/ROCm-Developer-Tools/HIP/pull/1425.  I will open up any permissions on my PR branch, if needed, and you can take it from here.

Reply:
Also, I ran the test code from https://github.com/ROCm-Developer-Tools/HIP/issues/1205 and it passed all tests.  I would consider that a more rigorous test case than this.  Should we consider adding it to the unit tests instead of this test?

Reply:
> 
> 
> Also, I ran the test code from #1205 and it passed all tests. I would consider that a more rigorous test case than this. Should we consider adding it to the unit tests instead of this test?

Thanks @jeffdaily . We can add new dtest. Could you please check this PR with SWDEV-212675 too if not already tried ?

Reply:
Hadn't tried to reproduce SWDEV-212675, but it is impossible to occur with this PR.  The difference between #1425 and this PR is that we do not insert foreign AQL packets into HCC's queue.  HCC's runtime ensures the underlying HSA queue will not overflow, but only if you use existing APIs (`create_marker`, etc).  I'm working on adapting the code from #1205 into a new test.